pandas/core/groupby/grouper.py

```diff
@@ -459,7 +459,7 @@ class Grouping:
       * groups : dict of {group -> label_list}
     """
 
-    _codes: np.ndarray | None = None
+    _codes: npt.NDArray[np.signedinteger] | None = None
```

Comment:
I'm not familiar with the code, just surprised that the integer-like types need to be signed. If it can be unsigned (`np.unsignedinteger`), you could use `np.integer` to combine both signed and unsigned integer-like types.

Reply:
all the relevant cases are signed, so this is as specific as the annotation can get i think